### ISSUE TITLE
Retrieve a valid set of codes after constraints resolution

### ISSUE THREAD

**Issue by dosse** [sdmx-twg/sdmx-rest#40]
Resolution of the constraints on the server side for big datasets that a client would not be able to process, for which he provides a restriction for one or more dimension (CubeRegion) and the service would then return the valid set of codes (partial codelist) from the other dimensions.

**Comment by dosse** [comment]
Potential new REST syntax:
`http://ws-entry-point/dataflow/AGENCY_ID/DATAFLOW_ID/latest?dynamicconstraint=CODE1+CODE2+CODE3..A&references=codelist&detail=resourcestubs`

New parameters used:
dynamicconstraint: specification of ad-hoc restriction of one or more dimensions re-using the key syntax for data
detail=resourcestubs: stubs are returned for resource artefacts but all available information is returned for the references (here the codelists)

This syntax doesn't allow for retrieving a particular partial codelist, as it would return all codelists. We would need another parameter to specify the requested codelist.

**Comment by tasosc** [comment]
Optionally we could also specify the component ID in order to retrieve only the codelist for that dimension:
`http://ws-entry-point/dataflow/AGENCY_ID/DATAFLOW_ID/latest/COMPONENT_ID?dynamicconstraint=CODE1+CODE2+CODE3..A&references=codelist&detail=resourcestubs`

**Comment by dosse** [comment]
MT implemented an extension of the SDMX Rest API with a new parameter option **detail=structureOnly**  in the data query for the SDMX-JSON representation. This solution is simple and sufficient for clients that can use SDMX-JSON because the SDMX-JSON data message contains all dimension codes for the available data. With **detail=structureOnly**, observations are not returned, only the structure information is included.

**Comment by tasosc** [comment]
Regarding the **detail=structureOnly**. What should the behavior be in other data formats e.g. XML StructureSpecificData ? 
For SDMX-JSON data and data formats in general could the existing **detail=serieskeysonly** be considered ?

**Comment by dosse** [comment]
The parameter detail=serieskeysonly returns the constraints twice, in the SDMX-JSON response: once as a Key Set (the data part) and once as a Cube Region (the structure part) and in the target use case only the Cube Region is needed. Receiving also the Key Set makes in many cases the response too large and unmanageable on client side. 

Xavier's comment with which I agree: "_I find the proposed implementation (of option **detail=structureOnly**) confusing. After all, it would mean that you perform a data query, although you don’t want data at all: You just need the cube region corresponding to a selection made by a user. Another indication in my opinion that the proposal is suboptimal is that it would work only with formats where data and metadata are combined (i.e. SDMX-JSON) but would not work with the other formats (SDMX-EDI, SDMX-ML, SDMX-CSV). This is in my opinion a sign of a suboptimal RESTful API design as, according to the RESTful principles, the URL should “work” regardless of the format selected for the representation._" Since this is a very important feature, I hope that a more optimal solution could be found.

**Comment by achronis** [comment]
In case it helps, from a technical point of view.

What is needed is a standard resource that could return on-the-fly/online/dynamic properties of the data, including the codes/values used and also other information like number of observations, and time period range without having to get the data itself (or even its series), since this could be impractical for technical reasons. 

For example for query for a partial codelist of component with codes that are used in the data we could have as input/output the following information:

**Query Input**

- Dataflow coordinates (agency, id, version). This is needed to resolve the dataset.
- The target dimension Id. This is needed to resolve the dimension's values within a dataset.
- Conditionally other dimension constraints. This could be used to restrict the set of target dimension values even more.
- Optionally the codelist coordinates (this can be resolved by the dataflow + target component information)

**Response**

- Structural metadata with a partial codelist. The codelist has only the codes that exist in the data for the requested dataflow and component. Codelist is needed instead of a e.g. a ContentConstraint because the codelist contains the names and descriptions of the codes. Although this is not optimal for un-coded dimensions. 

**Current suboptimal situation**

In case it helps, Eurostat's SDMX RI currently uses a SDMX v2.0 SOAP endpoint to achieve this. This is achieved using syntactically correct  Registry interface QueryStructureRequest/Response messages and by extending (a bit) the interpretation of the standard . Below is an example message

```xml
<RegistryInterface 
    xmlns="http://www.SDMX.org/resources/SDMXML/schemas/v2_0/message" 
    xmlns:common="http://www.SDMX.org/resources/SDMXML/schemas/v2_0/common" 
    xmlns:registry="http://www.SDMX.org/resources/SDMXML/schemas/v2_0/registry" 
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <Header>
        <ID>TEST</ID>
        <Test>true</Test>
        <Prepared>2010-04-03</Prepared>
        <Sender id="TEST"/>
    </Header>
    <QueryStructureRequest resolveReferences="false" returnDetails="true">
        <registry:CodelistRef>
            <registry:AgencyID>ESTAT</registry:AgencyID>
            <registry:CodelistID>CL_REF_AREA</registry:CodelistID>
            <registry:Version>1.0</registry:Version>
        </registry:CodelistRef>
        <registry:DataflowRef>
            <!-- Dataflow coordinates -->
            <registry:AgencyID>ESTAT</registry:AgencyID>
            <registry:DataflowID>SSTSIND_PROD_M</registry:DataflowID>
            <registry:Version>1.0</registry:Version>
            <!-- Constraint -->
            <registry:Constraint ConstraintType="Content">
                <common:ConstraintID>C222</common:ConstraintID>
                <common:CubeRegion isIncluded="true">
                    <common:Member isIncluded="true">
                        <!-- Constraint on FREQ dimension -->
                        <common:ComponentRef>FREQ</common:ComponentRef>
                        <common:MemberValue>
                            <common:Value>M</common:Value>
                        </common:MemberValue>
                    </common:Member>
                    <!-- Value-less component is the target component -->
                    <common:Member isIncluded="true">
                        <common:ComponentRef>REF_AREA</common:ComponentRef>
                    </common:Member>
                </common:CubeRegion>
            </registry:Constraint>
        </registry:DataflowRef>
    </QueryStructureRequest>
</RegistryInterface>
```

**Comment by sosna** [comment]
We will solve the use case as follows:
- There will be a new REST resource: `availableconstraint`
- The path parameters will be: `flowRef`, `key` and `providerRef` (like for data queries), as well as `componentId` (in case the client wants to retrieve the values for one dimension only).
- The query string parameters will be: `startPeriod`, `endPeriod`, `updatedAfter` (like for data queries), `references` (like for metadata queries), as well as mode (exact or available). 
- The output will be a `CubeRegion`. `ReferencePeriod` will be used for temporal coverage. 
- A dedicated `Annotation` will be used to express metrics (like the number of matching series or observations). Its type will be `sdmx_metrics`, its ID one of `series_count` or `obs_count` and the title will contain the measure.

**Comment by sosna** [comment]
Addressed in the upcoming v1.3.0 release
